### PR TITLE
[SYCL] Allow user control over range rounding, and disable for older SYCL spec.

### PR DESCRIPTION
**Pull request by rdeodhar**
This change lets users choose at compile-time whether to do range-rounding.
Rounding is always disabled when using older SYCL specs because the implementation creates const lambdas, which are incompatible with the mutable lambdas allowed in the old spec.
Signed-off-by: rdeodhar <rajiv.deodhar@intel.com>

### CODE REVIEW

**Comment by bader** [comment]
@rdeodhar, is this patch related to the https://github.com/intel/llvm/issues/3020?

**Comment by rdeodhar** [comment]
This patch is unrelated to #3020.

File: sycl/test/basic_tests/parallel_for_disable_range_roundup.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  int x;
+
+  x = 10;
+  try_rounding_off(x);
+
+  return 0;
+}
+
+// CHECK:  Run parallel_for with SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING
+// CHECK-NOT: parallel_for range adjusted from 10 to 32
```
Comment (romanovvlad) [inline sycl/test/basic_tests/parallel_for_disable_range_roundup.cpp:51]:
```suggestion
// CHECK-NOT: parallel_for range adjusted from 10
```
Suggest truncating the line to make the check more robust.
Another approach is to use LIT variables.

File: sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -sycl-std=2017 %s -o %t.out
+// RUN: env SYCL_PARALLEL_FOR_RANGE_ROUNDING_TRACE=1 %RUN_ON_HOST %t.out
+
+#include <CL/sycl.hpp>
+
+using namespace sycl;
+
+range<1> Range1 = {0};
+
+void check(const char *msg, size_t v, size_t ref) {
```
Comment (romanovvlad) [inline sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp:10]: Is it possible to merge the tests into one using `--check-prefix`?

File: sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -sycl-std=2017 %s -o %t.out
```
Comment (romanovvlad) [inline sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp:1]: Could you please add a check with different version of sycl-std?

File: sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  std::cout << msg << v << std::endl;
+  assert(v == ref);
+}
+
+int try_rounding_off(size_t size) {
+  range<1> Size{size};
+  int Counter = 0;
+  {
+    buffer<range<1>, 1> BufRange(&Range1, 1);
+    buffer<int, 1> BufCounter(&Counter, 1);
```
Comment (romanovvlad) [inline sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp:20]: It seems the tests need to be moved to https://github.com/intel/llvm-test-suite.

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp (hunk 4 above)
Comment (rdeodhar) [reply sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp:20]: Does this mean two separate PRs, one for the SYCL header change in intel/llvm and one for the tests in intel/llvm-test-suite? How to synchronize the two changes?

File: sycl/test/basic_tests/parallel_for_disable_range_roundup.cpp (hunk 1 above)
Comment (rdeodhar) [reply sycl/test/basic_tests/parallel_for_disable_range_roundup.cpp:51]: OK

File: sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp (hunk 4 above)
Comment (romanovvlad) [reply sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp:20]: @vladimirlaz Could you please help answering?

File: sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp (hunk 3 above)
Comment (rdeodhar) [reply sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp:1]: OK

File: sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp (hunk 4 above)
Comment (rdeodhar) [reply sycl/test/basic_tests/parallel_for_disable_range_roundup_syclstd.cpp]: Tests combined into one, and moved to test repo.

**Review by romanovvlad — APPROVED** [review]

**Comment by pvchupin** [comment]
@rdeodhar, please post link to new/updated tests in llvm-test-suite here. I missed this on merge.

**Comment by rdeodhar** [comment]
Yes, I am working on adding a test to llvm-test-suite.

**Pull request merged**